### ISSUE TITLE
admin page on webapp

### ISSUE THREAD

**Issue by prasadtalasila** [prasadtalasila/JavaAutolab#28]
We discussed the possibility of an admin page on the web application. The admin page would allow a course administrator to perform the following tasks.

Permit a course administrator to login using API key. We can debate the strategies of allowing third-party authentication vs local authentication vs API key authentication. API key-based authentication seems easiest to implement while providing the same-level of security.

Once an administrator logs in, client can use cookies to let administrator have continued access to /admin page until logout.

Admin is expected to perform the following operations.

1. Update labs.json file through a html editor.
2. Trigger a reevaluation request on an lab. We should debate the logic of triggering a reevaluation on an active lab. This is only for scenarios where the solution is updated during the lab itself.

When doing these things, we should keep in mind the future requirement of hosting multiple courses simultaneously.

@rajat503 can comment on all the suggested tasks.

**Comment by rajat503** [comment]
Keeping in mind the longer vision of hosting multiple courses:

**Login**: A password-based login system or Google OAuth (via bitsmail) should be a better way to go. We cannot expect all lab authors / course managers to go through the learning curve currently needed for Autolab. We cannot allow all course managers to SSH in machine 1 and API key restricts them to a single machine or a small set of machines (access on a CC terminal would be common for changes within the lab time). Cookies are a good idea but can be dangerous on public systems.

**labs.json**: I'd recommend putting a view with edit and add features, like a CRUD app. HTML editor for modifying JSON directly is very error prone. We can also consider moving labs.json to MySQL so that each course can have a separate table.

**Revaluation**: Most online judges ask for resubmission from the user if the test cases are updated during the contest. It would be better if students go through the submission process and see the scores for themselves than dynamically updating them. Refreshing the scoreboard can be considered provided we notify the users.

**Comment by rajat503** [comment]
We can also look at methods for giving students a detailed feedback for each commit (like we show after each submission) probably via e-mail once the SMTP is in place, when revaluation is done after the lab. This would make revaluation more transparent.

**Comment by prasadtalasila** [comment]
Login: OAuth requires us to put in a whitelist of all users and define roles for them. Seems too complex right now. Instead we can give API keys to the course admins. At a later point, we can put in the OAuth part, which can generate the session-specific API keys. The code developed with API would still work later on.

As far as cookies go, there is no other way to maintain an authenticated session in a client browser. We have to go through the headache of cookies. The best we can do is to delete the cookie on logout.

On adding support for multiple courses, I think fair bit of refactoring is required in all components of the application. Let's take it up later.

@tejas-sangol has already developed a HTML form-based editor for labs.json. It looks good and easy to use.

For us, resubmission on test case update is not really necessary. The code is already in Gitlab. We may optionally send an email to users; but, honestly that's a headache to users for short duration contests like labs. For long-duration contests where participants can do the work on and off, email update makes more sense. At best, we can make an email update an option; but, lets's postpone the e-mail option to a later point.
